### PR TITLE
Unify comment formatting inside the empty blocks

### PR DESCRIPTION
**Pull request by ah-yu**
Unify the format of comments inside the empty block 
```res
// before
let foo = [// hi
]

let foo = [//hi
/* hi */]

// now -- same as prettier
let foo = [
  // hi
]

let foo = [
  // hi
 /* hi */
]

```

### CODE REVIEW

File: tests/printer/comments/expected/typexpr.res.txt [hunk 1]
```diff
@@ -28,6 +28,10 @@ type t = /* c0 */ string /* c1 */ as 'x // after
 // Ptyp_poly
 type fn = {f: /* c0 */ 'a /* c1 */ 'b /* c2 */. /* c3 */ string /* c4 */}
 
+type fn = {/* comment1 */
```
Comment (IwanKaramazow) [inline tests/printer/comments/expected/typexpr.res.txt:31]: Can we tweak to output to look more like the input:
```
type fn = {
   /* comment1 */
   // comment2
 }
```

File: tests/printer/comments/expected/expr.res.txt [hunk 2]
```diff
@@ -107,6 +107,11 @@ let user = /* before */ {
   /* c4 */ age /* c5 */: /* c6 */ 31 /* c7 */,
 } // after
 
+// Pexp_record with empty record
+let user = /* before */ {/* comment1 */
```
Comment (IwanKaramazow) [inline tests/printer/comments/expected/expr.res.txt:111]: The comments inside look a bit suboptimal, can we go for:
```
{
   /* comment1 */
   // comment2
 }
```

File: src/res_comments_table.ml [hunk 3]
```diff
@@ -692,9 +692,12 @@ and walkTypeDeclaration (td : Parsetree.type_declaration) t comments =
     | Ptype_abstract | Ptype_open -> rest
     | Ptype_record labelDeclarations ->
       let () =
-        walkList
-          (labelDeclarations |> List.map (fun ld -> LabelDeclaration ld))
-          t rest
+        if List.length labelDeclarations = 0 then
```
Comment (cknitt) [inline src/res_comments_table.ml:695]: `List.length` iterates over the list unnecessarily, please use `match labelDeclarations with [] -> ...` instead.

**Comment by cknitt** [comment]
Also, I just fixed CI, please rebase on latest master to get CI to run.

File: tests/printer/comments/expected/jsx.res.txt [hunk 4]
```diff
@@ -14,9 +14,9 @@ module Cite = {
 
 <A /* comment */ />
 
-<A
-// Comment
-/>
+<A>
```
Comment (ah-yu) [inline tests/printer/comments/expected/jsx.res.txt:17]: Could we keep tags opening if these are comments inside? @cknitt 
I found this change was introduced in https://github.com/rescript-lang/syntax/commit/e913c4cf9881efa8d91778e72cc50a61ce651b78

Comment (cknitt) [reply tests/printer/comments/expected/jsx.res.txt:17]: Fine with me! That would make it easier to add code after the comment anyway.

**Comment by cristianoc** [comment]
Has the scope of this PR changed?
There seem to be differences outside empty records.

**Comment by ah-yu** [comment]
Yes. Now this PR is for adjusting the comments format inside an empty block(list, record, etc)

**Comment by cristianoc** [comment]
> Yes. Now this PR is for adjusting the comments format inside an empty block(list, record, etc)

Would you add an entry in the CHANGELOG.md as this is user-visible.
@IwanKaramazow any more comments?

**Review by IwanKaramazow — COMMENTED** [review]
If a user wrote `[/* a */]`, can we keep it on one line? Similar for the other "empty nodes".

We tend to look at the style of the author: did he write it over multiple lines or not?
* If it is on one line, keep it on one line (unless it breaks the column width).
* If it is written over multiple lines, force break it over multiple lines.

**Comment by ah-yu** [comment]
That sounds reasonable, will fix it later.

File: src/res_printer.ml [hunk 5]
```diff
@@ -705,19 +738,15 @@ and printModType ~customLayout modType cmtTbl =
...
-           [
-             Doc.lbrace;
-             Doc.indent
-               (Doc.concat
-                  [Doc.softLine; printCommentsInside cmtTbl modType.pmty_loc]);
-             Doc.softLine;
-             Doc.rbrace;
-           ])
+      let doc = printCommentsInside cmtTbl modType.pmty_loc in
+      if Doc.isNil doc then
```
Comment (cristianoc) [inline src/res_printer.ml:742]: This does the job but looks like tight coupling at a distance.
When is it that `printCommentsInside` currently produces `Nil`, and can we get hold of that information instead (e.g. return an optional).

**Review by cristianoc — APPROVED** [review]

File: src/res_printer.ml (hunk 5 above)
Comment (IwanKaramazow) [reply src/res_printer.ml:742]: Yes, in an ideal world, we shouldn't inspect the contents of a doc to make formatting decisions. The information we go off is the AST (+ comments).

@ah-yu is the `isNil` absolutely necessary?

Comment (ah-yu) [reply src/res_printer.ml:742]: `isNil` is only used in the development phase, then I could do a quick test without too many code changes.

Forgot to clean it up and will do it later.

File: src/res_printer.ml [hunk 6]
```diff
@@ -113,6 +113,12 @@ let hasNestedJsxOrMoreThanOneChild expr =
   in
   loop false expr
 
+let hasCommentsInside tbl loc =
+  match Hashtbl.find tbl.CommentTable.inside loc with
```
Comment (cristianoc) [inline src/res_printer.ml:117]: Can use find_opt

**Comment by cristianoc** [comment]
@ah-yu just ping when this is ready.

**Comment by ah-yu** [comment]
@cristianoc done

**Comment by cristianoc** [comment]
> @cristianoc done

Thanks a lot for contributing this!

**Pull request merged**